### ISSUE TITLE
Lost keyboard focus for new window when a layer surface (with exclusive kbd focus) is active

### ISSUE THREAD

**Issue by marcusbritanicus** [WayfireWM/wayfire#1204]
**Describe the bug**
Keyboard focus is not given to

**To Reproduce**
1. Open any window which takes keyboard focus (gtk/qt text editor may be: not vim)
2. Now open a layer surface that takes exclusive keyboard focus. (On-demand may also work, not tested)
    **Problem 1:** The layer surface now has keyboard focus, but the window which previously had the focus is still "marked" as active. SSD titlebar color and text cursor visible.
3. Now, open another window which takes keyboard focus (without closing the layer surface). This could be arranged by pressing a keybinding. 
     **Problem 2:** The newly opened window will be marked as "active" and will be "led to believe" it has keyboard focus - shows 
SSD titlebar color as active, and will show blinking text cursor.
4. Once the new window is open, close the layer surface.
     **Problem 3:** Both the windows "appear" to have keyboard focus - cursors blinking. SSD titlebar color is "active" for both windows.  See screenshot 1 (below).

**Expected behavior**
1. When the layer surface is opened, mark the window which previously had focus as "inactive"
2. When a new window is opened, it should be marked as "demands attention". Atleast it should not be marked as "active".
3. When the layer surface is closed, we have two options: 
  1. the newly opened window should get the keyboard focus and erase "demands attention"
  **OR**
  2. give the keyboard focus to the old window, and the newly opened window can continue to be in "demands attention" state till the user activates it with mouse or other means.

**Screenshots or stacktrace**
[Screenshot 1](https://imgur.com/a/mEhgWCg) - Observe that both windows appear to have focus.
[Screenshot 2](https://imgur.com/a/cvkvPf1) - This is one of the expected solutions.

**Wayfire version**
git

**Comment by ammen99** [comment]
> 2\. **Problem 1:** The layer surface now has keyboard focus, but the window which previously had the focus is still "marked" as active. SSD titlebar color and text cursor visible.

I don't see a problem with this, it is intentional behavior. When you click on your panel or something, you don't want the active window to fade away, right? The cursor showing is probably a bug of the application, it should have lost keyboard focus (via `wl_keyboard.leave`). This is what GTK applications do.


> * **Problem 2:** The newly opened window will be marked as "active" and will be "led to believe" it has keyboard focus - shows
>   SSD titlebar color as active, and will show blinking text cursor.

Yet again, I don't see this as a problem - it is supposed to be active, because it is. Active != has keyboard focus.

> 4\. **Problem 3:** Both the windows "appear" to have keyboard focus - cursors blinking. SSD titlebar color is "active" for both windows.  See screenshot 1 (below).

Now this is a real issue which needs fixing.

**Comment by marcusbritanicus** [comment]
> I don't see a problem with this, it is intentional behavior. When you click on your panel or something, you don't want the active window to fade away, right? The cursor showing is probably a bug of the application, it should have lost keyboard focus (via wl_keyboard.leave). This is what GTK applications do.

I agree with this in theory. Panels or some other similar surfaces, sure. How about a drop-down terminal? It is the "main window", and it is the active window. Currently, I don't see another other way to implement this other than as a layer surface. 
As for gtk applications, element is a gtk application (isn't it?) and the keyboard focus is still there. 

As for Qt Applications showing cursor, yes, I think it's a bug though where exactly I'm not certain. I get the following debug message, on **Step 3** (when the new window open when the layer surface is open with kbd focus):
```
Unexpected wl_keyboard.enter event. Keyboard already has focus
The compositor sent a wl_pointer.enter event before sending a leave event first, this is not allowed by the wayland protocol attempting to work around it by invalidating the current focus
```
Is it related to #1130 ?

> Yet again, I don't see this as a problem - it is supposed to be active, because it is. Active != has keyboard focus.

My bad. I might have misunderstood things a little bit. So how do we decide if a window is active? Also, what do we mean by keyboard focus?

**Comment by marcusbritanicus** [comment]
@ammen99 Thank you. I had not consciously noticed this... But yeah, the experience has been much more smooth.
